# Navigating the Single-Cell Landscape: A Comprehensive Workshop on scRNAseq

## Day 1: June 4th 2024
- **Where**: Karp 11 Conference Room
- **Course Material**:
    - Introduction to R/Rstudio
        - What is R studio and how to navigate it
        - How to work with R studio projects
        - Basics of R for scRNAseq
    - Basic understanding of the SC workflow
        - What are the key steps taken in scRNAseq data preprocessing 
        - What is the goal of data preprocessing
        - Why and how we need to normalize our data
        - Why do we need to select highly variable genes (HVG)
        - What is a K nearest neighbor graph
    - Normalization & HVG selection
        - How to normalize a scRNAseq count matrix
        - How do we select HVGs
    - Principal Component Analysis, Integration, Graph representation
        - How to carry out PCA
        - When and how do we need to integrate our data
        - Bulding a graph representation of our data

## Day 2: June 11th 2024
- **Where**: Karp 11 Conference Room
- **Course Material**:
    - Quality Control Analysis
        - Tips and tricks on how to carry out QC
        - What metrics do we need to look at and what do they mean
        - How can we detect low quality cells
        - How to spot doublets
    - Clustering
        - What is clustering and how does it work
        - What algorithms can we use
        - How do we find the right resolution
    - Differential expression and level 1 annotation
        - What is differential expression and how to do it
        - Evaluating the results
        - Annotating our clusters
    - Level 2 annotation and why we need it
        - Why do we need to subset and carry out a level-2 annotation
        - How to do it
