---
title: "A Beginner's Guide to scRNAseq - 2nd Edition"
author: "CDN team"
date: last-modified
date-format: "[Last compiled on] D MMMM, YYYY"
format:
  html:
    toc: true
    toc_float: true
    toc-location: left
    toc-depth: 4
    html-math-method: katex
    embed-resources: true
    self-contained-math: true
editor_options: 
  chunk_output_type: inline
---

## Day 1: Dec 3rd 2024

-   **Where**: Karp 6 Conference Room
-   **Course Material**:
    -   [Introduction to R/Rstudio/Seurat](https://github.com/CellDiscoveryNetwork/workshops/tree/main/beginners-guide-to-analyzing-scRNAseq-2nd-edition/prework) [[Recording]()]
        -   How to navigate Rstudio
        -   Fundamentals and data wrangling in R
        -   Working with a Seurat object
    -   [Basic sc-RNAseq workflow](https://github.com/CellDiscoveryNetwork/workshops/blob/main/beginners-guide-to-analyzing-scRNAseq-2nd-edition/day-1/The%20Basic%20scRNAseq%20Analysis%20Workflow.pdf) [[Recording]()]
        -   Main steps for scRNAseq workflow
        -   Intuition behind each step
    -   [Normalization & HVG selection](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/workshops/blob/main/beginners-guide-to-analyzing-scRNAseq-2nd-edition/day-1/3-norm-hvg.html) [[Recording]()]
        -   Why and how to normalize sc-RNAseq data
        -   Why and how to select highly variable genes
    -   [Quality Control](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/workshops/blob/main/beginners-guide-to-analyzing-scRNAseq-2nd-edition/day-1/4-QC.html) [[recording]()]
        -   How to carry out QC analysis
        -   What parameters do we need to look at and how to interpret them
        -   Computing doublet scores

## Day 2: Dec 10th 2024

-   **Where**: Karp 6 Conference Room
-   **Course Material**:
    -   [PCA, Integration & KNN graphs](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/workshops/blob/main/beginners-guide-to-analyzing-scRNAseq-2nd-edition/day-2/5-PCA_Harmony_kNN.html)[[Recording]()]
        -   Brief overview on PCA and how to use it for sc-RNAseq
        -   KNN-graph representation of the data
        -   Correcting batch effects with Harmony
        -   Tips and tricks during the QC process
    -   [Clustering](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/workshops/blob/main/beginners-guide-to-analyzing-scRNAseq-2nd-edition/day-2/6-Clustering.html) [[recording]()]
        -   How to cluster sc-RNAseq data
        -   What algorithms can we use
        -   How to assess if a clustering resolution is good
    -   [Differential Gene Expression & Level 1 Annotation](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/workshops/blob/main/beginners-guide-to-analyzing-scRNAseq-2nd-edition/day-2/7-dge-annotlvl1.html) [[recording]()]
        -   How to compute differentially expressed genes between clusters
        -   How to evaluate differential expression statistics
        -   Reference-based cell type annotation
        -   Manual cell type annotation
    -   [Subclustering & Level 2 Annotation](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/workshops/blob/main/beginners-guide-to-analyzing-scRNAseq-2nd-edition/day-2/8-Subclustering.html) [[recording]()]
        -   What does level 2 annotation mean and why we need it
        -   How to iteratively annotate a dataset

### Exit Forms
TBD
